server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/upload/InMemoryUploadRepository.java

```diff
@@ -0,0 +1,149 @@
+import org.apache.james.core.Username;
+import org.apache.james.jmap.api.exception.UploadNotFoundException;
+import org.apache.james.jmap.api.model.Upload;
+import org.apache.james.jmap.api.model.UploadId;
+import org.apache.james.jmap.api.model.UploadMetaData;
+import org.apache.james.jmap.api.upload.UploadRepository;
+import org.apache.james.mailbox.model.ContentType;
+import org.reactivestreams.Publisher;
+
+import com.google.common.base.Preconditions;
+
+import reactor.core.publisher.Mono;
+
+public class InMemoryUploadRepository implements UploadRepository {
+
+    private static final Map<UploadId, ImmutablePair<Username, BlobId>> uploadStore = new HashMap<>();
+    private static final Map<UploadId, UploadMetaData> uploadMetaDataStore = new HashMap<>();
+
+    static class UploadContext {
+        private final UploadId uploadId;
+        private final boolean exists;
+        private final byte[] data;
+
+        private UploadContext(UploadId uploadId,
+                              boolean exists,
+                              byte[] data) {
+            this.uploadId = uploadId;
+            this.exists = exists;
+            this.data = data;
+        }
+    }
+
+
+    private final BlobStore blobStore;
+    private final BucketName bucketName;
+
+    @Inject
+    public InMemoryUploadRepository(BlobStore blobStore) {
+        this.blobStore = blobStore;
+        this.bucketName = blobStore.getDefaultBucketName();
+    }
+
+    @Override
+    public Publisher<UploadId> upload(InputStream data, ContentType contentType, Username user) {
+        return Mono.just(checkUploadExits(data, user))
+            .flatMap(uploadContext -> {
+                if (!uploadContext.exists) {
+                    return store(uploadContext.data, uploadContext.uploadId, contentType, user);
+                } else {
+                    return Mono.just(uploadContext.uploadId);
+                }
+            });
+    }
+
+    @Override
+    public Publisher<Upload> retrieve(UploadId id, Username user) {
+        Preconditions.checkNotNull(id);
+        Preconditions.checkNotNull(user);
+
+        return retrieveBlobId(id, user)
+            .flatMap(blobId -> retrieveUpload(id, blobId));
+    }
+
+    private Mono<BlobId> retrieveBlobId(UploadId id, Username user) {
+        return Mono.justOrEmpty(uploadStore.get(id))
+            .filter(pair -> userHasAccessToUpload(user, pair.left))
+            .map(userAndBlobId -> userAndBlobId.right)
+            .switchIfEmpty(Mono.error(() -> new UploadNotFoundException(id)));
+    }
+
+    private Mono<Upload> retrieveUpload(UploadId id, BlobId blobId) {
+        return Mono.zip(Mono.from(blobStore.readBytes(bucketName, blobId)),
+            Mono.just(uploadMetaDataStore.get(id)))
+            .map(bytesAndMetadata -> Upload.from(bytesAndMetadata.getT2(), bytesAndMetadata.getT1()));
+    }
+
+    private boolean userHasAccessToUpload(Username userRequest, Username owner) {
```

Comment:
Is this method needed? Added value seems small to me...

Reply:
I thought we can extend logic in the future, so here is a reserve. 
I will implement checking logic in-line.

Reply:
I prefer having user.equals(owner) directly, I would find it clearer.

We can extract to a method later, when it makes sense IMO...